tracing/opencensus/http.go

```diff
@@ -0,0 +1,174 @@
+				trace.StringAttribute(ochttp.PathAttribute, req.URL.Path),
+				trace.StringAttribute(ochttp.UserAgentAttribute, req.UserAgent()),
+			)
+
+			if !cfg.public {
+				cfg.httpPropagate.SpanContextToRequest(span.SpanContext(), req)
+			}
+
+			return trace.NewContext(ctx, span)
+		},
+	)
+
+	clientAfter := kithttp.ClientAfter(
+		func(ctx context.Context, res *http.Response) context.Context {
+			if span := trace.FromContext(ctx); span != nil {
+				span.SetStatus(ochttp.TraceStatus(res.StatusCode, http.StatusText(res.StatusCode)))
+				span.AddAttributes(
+					trace.Int64Attribute(ochttp.StatusCodeAttribute, int64(res.StatusCode)),
+				)
+			}
+			return ctx
+		},
+	)
+
+	clientFinalizer := kithttp.ClientFinalizer(
+		func(ctx context.Context, err error) {
+			if span := trace.FromContext(ctx); span != nil {
+				if err != nil {
+					span.SetStatus(trace.Status{
+						Code:    trace.StatusCodeUnknown,
+						Message: err.Error(),
+					})
+				}
+				span.End()
+			}
+		},
+	)
+
+	return func(c *kithttp.Client) {
+		clientBefore(c)
+		clientAfter(c)
+		clientFinalizer(c)
+	}
+}
+
+// HTTPServerTrace enables OpenCensus tracing of a Go kit HTTP transport server.
+func HTTPServerTrace(options ...TracerOption) kithttp.ServerOption {
+	cfg := TracerOptions{
+		sampler:       trace.AlwaysSample(),
+		httpPropagate: &b3.HTTPFormat{},
+	}
+
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	serverBefore := kithttp.ServerBefore(
+		func(ctx context.Context, req *http.Request) context.Context {
+			var (
+				spanContext trace.SpanContext
+				span        *trace.Span
+				name        string
+				ok          bool
+			)
+
+			if cfg.name != "" {
+				name = cfg.name
+			} else {
+				name = req.Method + " " + req.URL.Path
+			}
+
+			spanContext, ok = cfg.httpPropagate.SpanContextFromRequest(req)
+			if ok && !cfg.public {
+				ctx, span = trace.StartSpanWithRemoteParent(
+					ctx,
+					name,
+					spanContext,
+					trace.WithSpanKind(trace.SpanKindServer),
+					trace.WithSampler(cfg.sampler),
+				)
+			} else {
+				ctx, span = trace.StartSpan(
+					ctx,
+					name,
+					trace.WithSpanKind(trace.SpanKindServer),
+					trace.WithSampler(cfg.sampler),
+				)
+				if ok {
+					span.AddLink(trace.Link{
+						TraceID:    spanContext.TraceID,
+						SpanID:     spanContext.SpanID,
+						Type:       trace.LinkTypeChild,
+						Attributes: nil,
+					})
+				}
+			}
+
+			span.AddAttributes(
+				trace.StringAttribute(ochttp.MethodAttribute, req.Method),
+				trace.StringAttribute(ochttp.PathAttribute, req.URL.Path),
+			)
+
+			return ctx
+		},
+	)
+
+	serverFinalizer := kithttp.ServerFinalizer(
+		func(ctx context.Context, code int, r *http.Request) {
+			if span := trace.FromContext(ctx); span != nil {
+				span.SetStatus(ochttp.TraceStatus(code, http.StatusText(code)))
+
+				if rs, ok := ctx.Value(kithttp.ContextKeyResponseSize).(int64); ok {
+					span.AddAttributes(
+						trace.Int64Attribute("http.response_size", rs),
```

Comment:
Why is response size "special" in this way? Can other attributes be special, too?

Reply:
I'm not sure what you're noticing as special here. Can you explain?

Reply:
Sorry, that _was_ a bit obtuse. I guess your code here responds to [these lines of code here](https://github.com/go-kit/kit/blob/master/transport/http/server.go#L90-L91), and I had mistakenly thought we were adding more to the context than we actually were. Disregard.